Deserialization refactor: introduce new deserialization traits

wprzytula opened the pull request:
Fixes #961
Fixes #963

Huge credits to @piodul for creating most of the new deserialization framework!

## The new traits
This PR introduces the new deserialization traits, `DeserializeValue` and `DeserializeRow`, which are going to replace `FromCqlVal` and `FromRow`, respectively.

If a type implements `DeserializeValue<'f>`, this means that it can take a buffer of lifetime `'f` and deserialize itself, treating the contents of the buffer as a CQL value. Analogously, `DeserializeRow<'f>` allows for deserialization of types that are supposed to represent a whole row returned in the results.

Deserialization is now split into two phases: type checking and actual deserialization. Both traits have two methods: `type_check` and `deserialize`. The idea behind this is to validate the column types of the response only once, after the response is received. The `deserialize` method is then called for each row/value to perform the actual parsing: that method is allowed to assume that `type_check` was called and may skip some type checking for performance (although it cannot use that assumption to perform unsafe operations, it is not an unsafe method).
    
The previous framework only supported deserialization of types that exclusively own their data, for example `String` or `Vec<u8>`. However, the new framework allows for types that borrow or co-own the frame buffer:
    - Types that borrow the buffer are `&'f str` and `&'f [u8]`. They just point to the raw data from the serialized response. Rust's lifetime system  makes sure that the the user doesn't deallocate the serialized  response until using the deserialized types that point to the buffer are dropped. Apart from the aforementioned types, a bunch of iterator types have been introduced that allow consuming collections without allocations.
    - Serialized frame is represented with `bytes::Bytes`. It is possible to create a subslice (also of type `Bytes`) that keep the original alive. The type being deserialized can obtain access to such subslice. Like in the case of e.g. `&str` vs. `String`, keeping `Bytes`  allows to avoid an allocation, but it is also easier to handle as deserialized type doesn't have to be bound by any lifetime. It important to be careful when handling such types as they keep whole serialized frame alive, even if they only point to a subslice. This can lead to a space leak and more memory being used than necessary.

## Implementation of the new traits for all supported types

**All the previously supported types** (i.e. those that could be deserialized using the old framework) **are supported now as well**.
There is only one exception: CQL List is no longer deserializable into a set (`HashSet` or `BTreeSet`), because such conversion could be lossy, which is an unwanted property for deserialization.

## Errors

The error framework is analogous to one used in the serialization framework: `DeserializationError` is added that resembles `SerializationError` and is type-erased as well.

As the new deserialization framework has separate `type_check()` and `deserialize()` function (contrary to the serialization framework, which has only one `serialize()` method), it makes sense for them to return distinct errors.
That's why a new `TypeCheckError` is added, too, which is type-erased as well.

## Testing

- All new types are unit tested. 
- Additionally, a decent test suite is added to verify compatibility of the old and the new framework. As in a later commit the old framework (`deser_cql_value`) is refactored to use the new one, the compatibility tests are then removed.
- Another test suite is added to check that serialization composed with deserialization yields identity.
- New errors are tested, too, although not exhaustively. I'd prefer not to bog down too much in testing them thoroughly.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -0,0 +1,166 @@
+
+    /// Returns a reference to the slice.
+    #[inline]
+    pub fn as_slice(&self) -> &'frame [u8] {
+        self.frame_subslice
+    }
+
+    /// Returns `true` if the slice has length of 0.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.frame_subslice.is_empty()
+    }
+
+    /// Returns a reference to the Bytes object which encompasses the slice.
+    ///
+    /// The Bytes object will usually be larger than the slice returned by
+    /// [FrameSlice::as_slice]. If you wish to obtain a new Bytes object that
+    /// points only to the subslice represented by the FrameSlice object,
+    /// see [FrameSlice::to_bytes].
+    #[inline]
+    pub fn as_bytes_ref(&self) -> &'frame Bytes {
+        self.original_frame
+    }
+
+    /// Returns a new Bytes object which is a subslice of the original slice
+    /// object.
+    #[inline]
+    pub fn to_bytes(&self) -> Bytes {
+        self.original_frame.slice_ref(self.frame_subslice)
+    }
```
### Why we need a slice
The reason why we need to store a slice is that we want a lifetime that is longer than one obtained when coercing `Bytes` to a slice in the body of a function. That is, we want to allow deserializing types that borrow from the frame, which resides in `QueryResult`.
Consider a function with the signature:
```rust
fn fun(b: Bytes) { ... }
```
This function cannot return a type that borrows from the frame, because any slice created from `b` inside `fun` cannot escape `fun`.
Conversely, if a function has signature:
```rust
fn fun(s: &'frame [u8]) { ... }
```
then it can happily return types with lifetime `'frame`.

### Why we need the full frame?
We don't. We only need to be able to return `Bytes` encompassing our subslice. However, the design choice was made to only store a reference to the original `Bytes` object residing in `QueryResult`, so that we avoid cloning `Bytes` when performing subslicing on `FrameSlice`. We delay the `Bytes` cloning, normally a moderately expensive operation involving cloning an `Arc`, up until it is really needed.

### Why not different design?
- why not a `&'frame [u8]` only? Because we won't to enable deserializing types containing owned `Bytes`, too.
- why not a `Bytes` only? Because we need to propagate the `'frame` lifetime.
- why not a `&'frame Bytes` only? Because we want to somehow represent subslices, and subslicing `&'frame Bytes` return `Bytes`, not `&'frame Bytes`.

Is the above convincing enough? Please hint me which of the above would you like to be put in the comment explanation.

wprzytula commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -0,0 +1,101 @@
+/// Represents a raw, unparsed column value.
+#[non_exhaustive]
+pub struct RawColumn<'frame> {
+    pub spec: &'frame ColumnSpec,
```
This is how @piodul designed the framework.

piodul replied:
With prepared statement metadata caching IMO it's not worth making the column specs lazily deserializable.

Lorak-mmk replied:
Metadata caching is disable by default because of the issue with `SELECT *`, isn't it?

muzarski replied:
> Metadata caching is disable by default because of the issue with `SELECT *`, isn't it?

It is. I decided to disable it by default for two reasons:
1. rust driver issue - we decided to leave metadata immutable. Performance-aware users should not alter the schema after enabling this optimisation.
2. cql protocol issue - the multi client scenario (related to `SELECT *`), where only one client reprepares the statement. The other clients are not notified about the repreparation and so they store outdated metadata. To reproduce this issue, the rust driver issue mentioned above needs to be solved firstly.

wprzytula replied:
> It is. I decided to disable it by default for two reasons:

    rust driver issue - we decided to leave metadata immutable. Performance-aware users should not alter the schema after enabling this optimisation.
    cql protocol issue - the multi client scenario (related to SELECT *), where only one client reprepares the statement. The other clients are not notified about the repreparation and so they store outdated metadata. To reproduce this issue, the rust driver issue mentioned above needs to be solved firstly (because, as of now, the driver doesn't update the metadata anyway).

This should definitely be put in the docstring of the optimisation setting, so that it is clear both for us and for users.

muzarski replied:
> > It is. I decided to disable it by default for two reasons:
> 
> > 1. rust driver issue - we decided to leave metadata immutable. Performance-aware users should not alter the schema after enabling this optimisation.
> > 2. cql protocol issue - the multi client scenario (related to SELECT *), where only one client reprepares the statement. The other clients are not notified about the repreparation and so they store outdated metadata. To reproduce this issue, the rust driver issue mentioned above needs to be solved firstly (because, as of now, the driver doesn't update the metadata anyway).
> 
> This should definitely be put in the docstring of the optimisation setting, so that it is clear both for us and for users.

I agree. I can take care of it.

wprzytula commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -0,0 +1,101 @@
+where
+    Self: Sized,
+{
+    /// Checks that the schema of the result matches what this type expects.
+    ///
+    /// This function can check whether column types and names match the
+    /// expectations.
+    fn type_check(specs: &[ColumnSpec]) -> Result<(), ParseError>;
```
Can you see any usecase for trait objects in deserialization? I can't think of any.

wprzytula replied:
row's `type_check()` is going to be called exactly once per response, before any row is deserialized.
OTOH row's `deserialize()` is called once per each row, so if we only type check once, this is a win.

wprzytula replied:
`DeserializeRow`  and `DeserializeCql` are both immanently not object-safe because they [violate the rule of not having `Sized` as their supertrait](https://doc.rust-lang.org/reference/items/traits.html#object-safety).

wprzytula commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -0,0 +1,101 @@
+    fn deserialize(row: ColumnIterator<'frame>) -> Result<Self, ParseError>;
+}
+
+impl<'frame> DeserializeRow<'frame> for ColumnIterator<'frame> {
+    #[inline]
+    fn type_check(_specs: &[ColumnSpec]) -> Result<(), ParseError> {
+        Ok(())
+    }
+
+    #[inline]
+    fn deserialize(row: ColumnIterator<'frame>) -> Result<Self, ParseError> {
+        Ok(row)
+    }
+}
```
See https://github.com/scylladb/scylla-rust-driver/pull/665/commits/10eceb30868711aa12dde702ad53fa18dd999630#r1153375481
I will put the explanation in a comment.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -0,0 +1,35 @@
+/// For tips on how to write a custom implementation of this trait, see the
+/// documentation of the parent module.
+///
+/// The crate also provides a derive macro which allows to automatically
+/// implement the trait for a custom type. For more details on what the macro
+/// is capable of, see its documentation.
```
Right, but quoting @piodul:

> It is very common for derive macros to have exactly the same name as the trait they implement. To me at least, it's confusing when you write #[derive(SerializeUdt)] while there is no corresponding SerializeUdt trait.

Do you suggest that `{Des,S}erializeCql` be derivable with `#[derive({Des,S}erializeUdt]` ?

wprzytula replied:
I'll change `DeserializeCql` to `DeserializeValue`, to align with #905.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -32,4 +32,78 @@ where
+        T: Debug + PartialEq,
+    {
+        let raw = serialize(typ, val);
+        compat_check::<T>(typ, raw);
```
We could, but is this going to be beneficial here?
If I were to add such tests, I would put them higher in the module tree, as "integration tests" between serialization and deserialization.

Lorak-mmk replied:
Higher up is fine

Lorak-mmk replied:
I think such tests would be valuable, but no need to implement them if it's too much work.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -217,6 +217,49 @@ impl_strict_type!(
+    }
+);
+
+// TODO: Deserialization for string::String<Bytes>
```
See https://github.com/scylladb/scylla-rust-driver/pull/665/commits/10eceb30868711aa12dde702ad53fa18dd999630#r1153450244

wprzytula replied:
https://crates.io/crates/string

Lorak-mmk replied:
> See [10eceb3#r1153450244](https://github.com/scylladb/scylla-rust-driver/commit/10eceb30868711aa12dde702ad53fa18dd999630#r1153450244)

What is this link supposed to point to? It points me to some commit by piodul, but I see no mention of `string::String` there



> https://crates.io/crates/string

Do we support deserialization of this now? If not, then let's leave it as TODO, but if we do then we shouldn't make a regression on this

wprzytula replied:
We don't support it yet.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -66,6 +66,75 @@ where
+                    _ => Err(ParseError::BadIncomingData(format!(
+                        "Expected {}, got {:?}",
+                        $cql_name, typ,
+                    ))),
```
I'll take a look at that machinery.

wprzytula replied:
I'm refactoring the code to use machinery similar to that used in serialization framework.

wprzytula replied:
Done!

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -833,6 +862,18 @@ mod tests {
         compat_check::<IpAddr>(&ColumnType::Inet, make_ip_address(ipv4));
         compat_check::<IpAddr>(&ColumnType::Inet, make_ip_address(ipv6));
+
+        // uuid and timeuuid
+        // new_v4 generates random UUIDs, so these are different cases
+        let uuid1 = Uuid::new_v4();
+        let uuid2 = Uuid::new_v4();
+        let uuid3 = Uuid::new_v4();
+        compat_check_serialized::<Uuid>(&ColumnType::Uuid, &uuid1);
+        compat_check_serialized::<Uuid>(&ColumnType::Uuid, &uuid2);
+        compat_check_serialized::<Uuid>(&ColumnType::Uuid, &uuid3);
+        compat_check_serialized::<CqlTimeuuid>(&ColumnType::Timeuuid, &CqlTimeuuid::from(uuid1));
+        compat_check_serialized::<CqlTimeuuid>(&ColumnType::Timeuuid, &CqlTimeuuid::from(uuid2));
+        compat_check_serialized::<CqlTimeuuid>(&ColumnType::Timeuuid, &CqlTimeuuid::from(uuid3));
```
IDK, that's what I got.
I can make it a loop.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -521,6 +521,44 @@ impl_strict_type!(
+/// `MaybeEmpty` was introduced to help support this quirk for Rust types
+/// which can't represent the empty, additional value.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
+pub enum MaybeEmpty<T> {
+    Empty,
+    Value(T),
+}
```
Will check and answer.

See also https://github.com/scylladb/scylla-rust-driver/pull/665/commits/10eceb30868711aa12dde702ad53fa18dd999630#r1161057969

wprzytula replied:
This is the legacy way of deserializing Empty values:

```rust
// Implement from_cql<Option<CqlValue>> for Option<T> for every type that has from_cql<CqlValue>
// Value inside Option gets mapped from CqlValue to T
impl<T: FromCqlVal<CqlValue>> FromCqlVal<Option<CqlValue>> for Option<T> {
    fn from_cql(cql_val_opt: Option<CqlValue>) -> Result<Self, FromCqlValError> {
        match cql_val_opt {
            Some(CqlValue::Empty) => Ok(None),
            Some(cql_val) => Ok(Some(T::from_cql(cql_val)?)),
            None => Ok(None),
        }
    }
}
```

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -860,6 +860,61 @@ where
+// tuples
+
+// Implements tuple deserialization.
+// The generated impl expects that the serialized data will contain at least
```
See https://github.com/scylladb/scylla-rust-driver/pull/665/commits/10eceb30868711aa12dde702ad53fa18dd999630#r1174279582
It seems that the comment hasn't been fixed yet.

wprzytula replied:
Deserializing tuples to `Vec<CqlValue> unfortunately conflicts with the blanket implementation for deserializing lists and sets into `Vec<T>`. I don't see a solution to this conflict.

Lorak-mmk replied:
> See [10eceb3#r1174279582](https://github.com/scylladb/scylla-rust-driver/commit/10eceb30868711aa12dde702ad53fa18dd999630#r1174279582) It seems that the comment hasn't been fixed yet.

This points me to the commit, did you mean to mention some specific part of it?

Lorak-mmk replied:
> Deserializing tuples to `Vec<CqlValue>` unfortunately conflicts with the blanket implementation for deserializing lists and sets into `Vec<T>`. I don't see a solution to this conflict.

It's OK, I think tupes can still be deserialized to just `CqlValue` so it doesn't matter

wprzytula replied:
> This points me to the commit, did you mean to mention some specific part of it?

There had already been a same remark and Piodul answered that the comment was wrong. I've fixed it.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -242,6 +292,69 @@ mod tests {
+            compat_check_serialized::<BigDecimal>(&ColumnType::Decimal, &num1);
+            compat_check_serialized::<BigDecimal>(&ColumnType::Decimal, &num2);
+            compat_check_serialized::<BigDecimal>(&ColumnType::Decimal, &num3);
+        }
```
No idea.

piodul replied:
> Why some types are only tested in this function and some are both tested here and in a separate function?

I'm not entirely sure because it was a long time ago, so what I'll write here might be wrong (with more probability than usual, at least).

There are two kinds of tests in this file:

1. Tests which verify that the `DeserializeCql` works properly (this is new code which needs to be tested)
2. Tests which verify that the `DeserializeCql` implementation agrees with the legacy `FromCqlVal`-based one

They both serve a different purpose.

Lorak-mmk replied:
My main question still remains: why are some types only tested in the second way and not the first way?

piodul replied:
I don't remember, sorry. I might have just forgot to write the first kind of test for some types.

Lorak-mmk replied:
Ok, unless there is a reason for some types to be absent from one type of tests please implenent both @wprzytula

wprzytula commented on scylla-cql/src/types/deserialize/mod.rs:
````diff
@@ -1,3 +1,155 @@
+//! }
+//! ```
+//!
+//! ## Reference-counted types (`DeserializeCql` only)
````
@piodul Do you have any idea for this? I don't unfortunately.

piodul replied:
Not sure. This is mostly relevant to `DeserializeCql` implementors as it is up to them to choose the way they refer to the original data, `DeserializeRow` implementors only aggregate `DeserializeCql` implementors.

Perhaps we can drop the mention of `DeserializeCql`.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -66,6 +66,75 @@ where
+            fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+                // TODO: Format the CQL type names in the same notation
+                // that ScyllaDB/Casssandra uses internally and include them
+                // in such form in the error message
```
What notation do you have in mind? Is there some list of those types in the notation you and the comment mention?

Lorak-mmk replied:
I'm not sure what notation the comment has in mind, as it is not my PR - I just think that if this is something that should be done here then it should probably be done in serialization too because it also formats CQL types.

wprzytula replied:
@piodul Could you shed some light here?

piodul replied:
I think the "notation used internally" was just supposed to mean "CQL syntax". I think it's a good UX if the error messages display the types in CQL syntax.

wprzytula replied:
Is the  syntax currently displayed in the code in line with the CQL syntax (as you mean it)?

piodul replied:
I have no idea :) I didn't touch the rust driver code for a while. You can write an example that triggers this error and see for yourself.

wprzytula replied:
So from what I can see, with simple types (non-collections) our naming differs only in being non-lowercase. In case of collections, though, it may be a bit more complicated. I'll investigate it further, but is it really a big deal to have the types displayed in the same case as the specification?

piodul replied:
I don't think this is a big deal as long as it is possible to understand which CQL type the error message is about. It's definitely not something that would block the merge, I think that's why I left a todo there.

However, from the UX perspective, I think it's worth it - I suppose that the CQL syntax will be most familiar to users, so it would be nice to implement this at some point.

wprzytula replied:
OK, I consider this out of scope of this PR. In the future we could make a code-wide change to use CQL syntax everywhere.

wprzytula commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -0,0 +1,166 @@
+        assert_eq!(subslice1.as_slice(), CELL1);
+        assert_eq!(subslice2.as_slice(), CELL2);
+
+        assert_eq!(
+            subslice1.as_bytes_ref() as *const Bytes,
+            &frame as *const Bytes
+        );
+        assert_eq!(
+            subslice2.as_bytes_ref() as *const Bytes,
+            &frame as *const Bytes
+        );
```
> You can use https://doc.rust-lang.org/std/ptr/fn.eq.html - then you won't have to cast to pointer manually.

But then I won't be able to use `assert_eq!`, which provides nice error message showing both values. I prefer to stick to the current version.

wprzytula replied:
IMO raw pointers are perfectly OK unless you dereference them; OTOH custom `assert!` requires needless verbosity and functionality duplication from `assert_eq!`.

wprzytula commented:
v2:
- addressed most review comments (though `DeserializeCql` -> `DeserializeValue` rename has not yet been done),
- set up and used an error framework similar to that used in serialization.

While reviewing, please pay attention to the error types used, especially to the use of `GenericParseError` variant (which I hate, but before errors refactor I can do little about this).

muzarski commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -3,7 +3,7 @@
 
 #[cfg(feature = "chrono")]
-use chrono::{NaiveDate, NaiveTime};
+use chrono::{DateTime, NaiveDate, NaiveTime, TimeZone as _, Utc};
```
I'd remove the `use` clause from here. We already use full paths for `time` types. Let's be consistent and do the same for `chrono` types as well. This will be eventually removed once we decide to support other `chrono` versions.

muzarski reviewed (changes requested):
Additional question: do we want to expose all of the iterator types to the users? I'm aware that `UdtIterator` might be useful to implement deserialization for custom user types. What about other iterators such as `ListLikeIterator` or `MapIterator`? Do you see any reason for exposing these?

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -573,6 +575,26 @@ impl_emptiable_strict_type!(
+        } else {
+            Err(mk_deser_err::<Self>(
+                typ,
+                BuiltinDeserializationErrorKind::ValueOverflow,
```
nit: The context you marked (4 lines showed here) aren't enough for me to quickly know what code you are referring to. I'll have to check it using line numbers. For readability it's better to grasp more LoC in the spoiler.

wprzytula replied:
Line numbers weren't enough, because they changed across commits...
I had to look through all commits until I found this comment.

wprzytula replied:
Added a dedicated variant for this error case.

wprzytula commented:
> Additional question: do we want to expose all of the iterator types to the users? I'm aware that `UdtIterator` might be useful to implement deserialization for custom user types. What about other iterators such as `ListLikeIterator` or `MapIterator`? Do you see any reason for exposing these?

1) exposing them might be useful for cpp-rust-driver (but in this aspect it's better to start with private visilibity and make it public only if really needed there),
2) it's not `scylla-cql` crate that should control visibility of the exports; it's `scylla` crate's responsibility to organise exports, just like you've recently done wrt some other exports from `scylla-cql`. Ideally, **only leaf items** (e.g. structs, consts, enums, etc.) are re-exported by `scylla` from `scylla-cql`, no node items (i.e. modules) - this gives the finest control over visibility.

wprzytula commented:
v3:
    - addressed more review comments, including: better names for `FrameSlice` methods.
    - still TODO: DeserializeCql -> DeserializeValue, keep `BuiltInTypeCheckError` separate for serialization and deserialization for more accurate error messages.

wprzytula commented:
Rebased on main.

wprzytula commented:
v3.1:
- replaced `assert!(matches!(...))` with `assert_matches!(...)` codewide.

wprzytula commented:
v3.2:
- rebased on main to fix clippy complaints.
- an independent prefix of this PR has been sent as #1004 to make this PR a bit smaller.

wprzytula commented:
v4:
- type check errors are no longer shared between serialization and deserialization. Entirely different types for `BuiltinTypeCheckError` are introduced for deserialization.

wprzytula commented:
v4.1:
- fully qualify `chrono` types

wprzytula commented:
v4.2 - fixed outdated docstring that led to CI failing.

wprzytula commented:
Reviewers: is it acceptable for you that DeserializeCql is only renamed to DeserializeValue in one commit at the end of this PR? This saves me a lot of time to rename it in all consecutive commits.

muzarski reviewed (changes requested):
I've got a general comment about `DeserializationError`. The docstring mentions:
```
/// As for the impls defined or generated
/// by the driver itself, the following errors can be returned:
/// - [`row::BuiltinDeserializationError`]
/// - [`value::BuiltinDeserializationError`]
```

In my opinion, that's not entirely true, since there are places where nested `DeserializationError`s are created directly from `ParseError` (see the `Deserialization::new()` usages).

muzarski commented on scylla-cql/src/types/deserialize/result.rs:
```diff
@@ -70,12 +71,72 @@ impl<'frame> Iterator for RowIterator<'frame> {
+    fn next(&mut self) -> Option<Self::Item> {
+        let raw = match self.inner.next() {
+            Some(Ok(raw)) => raw,
+            Some(Err(err)) => return Some(Err(DeserializationError::new(err))),
```
Sorry for the context (it's too small to recognize where is the issue). It's the implementation of `Iterator` for `TypedRowIterator`.

wprzytula replied:
Fixed, together with making the syntax more consise and clean using FP.

wprzytula commented:
> I've got a general comment about `DeserializationError`. The docstring mentions:
> 
> ```
> /// As for the impls defined or generated
> /// by the driver itself, the following errors can be returned:
> /// - [`row::BuiltinDeserializationError`]
> /// - [`value::BuiltinDeserializationError`]
> ```
> 
> In my opinion, that's not entirely true, since there are places where nested `DeserializationError`s are created directly from `ParseError` (see the `DeserializationError::new()` usages).

I'm glad you noticed that. However, note that all `DeserializationError::new()` call results are placed _**inside**_ another `BuiltinDeserializationError` (can you verify that together with me?). This means that if the user hold a `DeserializationError` produces by our lib code, then the outermost error there will be always `BuiltinDeserializationError`. Am I right?

muzarski commented:
> > I've got a general comment about `DeserializationError`. The docstring mentions:
> > ```
> > /// As for the impls defined or generated
> > /// by the driver itself, the following errors can be returned:
> > /// - [`row::BuiltinDeserializationError`]
> > /// - [`value::BuiltinDeserializationError`]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In my opinion, that's not entirely true, since there are places where nested `DeserializationError`s are created directly from `ParseError` (see the `DeserializationError::new()` usages).
> 
> I'm glad you noticed that. However, note that all `DeserializationError::new()` call results are placed _**inside**_ another `BuiltinDeserializationError` (can you verify that together with me?). This means that if the user hold a `DeserializationError` produces by our lib code, then the outermost error there will be always `BuiltinDeserializationError`. Am I right?

Yes, you are right. That's why I mentioned that only nested errors are created in a way that is not consistent with the docs. I don't think we need to adjust the docs or internal usages of `DeserializationError::new()` in any way. I just pointed out a small inconsistency, to make sure that we all are aware of it and that it does not affect users.

wprzytula commented:
> Reviewers: is it acceptable for you that `DeserializeCql` is only renamed to `DeserializeValue` in one commit at the end of this PR? This saves me a lot of time to rename it in all consecutive commits.

Fixed this. Commits now mention `DeserializeValue` from the very first place.

Lorak-mmk commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -23,9 +187,9 @@ use thiserror::Error;
+///   returned from [`DeserializeValue::type_check`](value::DeserializeValue::deserialize)
 ///   instead both in the case of builtin impls and impls generated by the
-///   `DeserializeCql` macro. It won't be returned by the `Session` directly,
+///   `DeserializeValue` macro. It won't be returned by the `Session` directly,
```
This should me squashed into the commit that introduced it.

Lorak-mmk commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -49,9 +213,9 @@ impl TypeCheckError {
 ///   instead both in the case of builtin impls and impls generated by the
-///   `DeserializeCql` macro. It won't be returned by the `Session` directly,
+///   `DeserializeValue` macro. It won't be returned by the `Session` directly,
 ///   but it might be nested in the [`row::BuiltinDeserializationError`].
```
ditto

Lorak-mmk commented on scylla-cql/src/types/deserialize/mod.rs:
Commit message for commit `scylla-cql/types: introduce new deserialization traits ` mentions `DeserializeCql` instead of `DeserializeValue`

Lorak-mmk commented on scylla-cql/src/types/deserialize/frame_slice.rs:
```diff
@@ -0,0 +1,188 @@
+// cloning Bytes when performing subslicing on FrameSlice. We delay the Bytes cloning, normally a moderately
+// expensive operation involving cloning an Arc, up until it is really needed.
+//
+// ## Why not different design
+//
+//      - why not a &'frame [u8] only? Because we want to enable deserializing types containing owned Bytes, too.
+//      - why not a Bytes only? Because we need to propagate the 'frame lifetime.
+//      - why not a &'frame Bytes only? Because we want to somehow represent subslices, and subslicing
+//        &'frame Bytes return Bytes, not &'frame Bytes.
+#[derive(Clone, Copy, Debug)]
```
Another option worth mentioning here would be to use unsafe: you could store only Bytes and transmute the lifetime of the returned slice.

wprzytula replied:
I think it is agreed upon that we don't allow any `unsafe` code in the driver itself.
Btw, we could enforce it by using `#![forbid(unsafe_code)` crate-wide. WDYT?

Lorak-mmk replied:
I know we don't use unsafe, that's why I only want to mention it as an alternative, not implement it.

We can enforce it, but I don't think it's important, I doubt any unsafe can get through review unnoticed.

Lorak-mmk commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -114,10 +115,97 @@ fn mk_deser_err_named(
+    /// A generic deserialization failure - legacy error type.
+    GenericParseError(ParseError),
+}
 
```
Is this only used for tests? Maybe it's changed in later commits, but I see it is introduced for tests right now.
If so, shouldn't it be a different error as this one is useless for the user? I thought that was the idea behind `DeserializationError` being dynamic: built-in errors only need to cover errors that may be returned by default implementations, and different implementations (like tests) can return different errors.

wprzytula replied:
The problem is that we we want to have a `BuiltinDeserializationError` for proper printing of `rust_name` and `got`, but for this we need to have a concrete `BuiltinDeserializationErrorKind` in hand. To achieve this in cases when we have a `ParseError` in hand, the `GenericParseError` variant is introduced.

Lorak-mmk commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -36,6 +36,40 @@ where
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, DeserializationError> {
+        let mut val = ensure_not_null_slice::<Self>(typ, v)?;
+        let cql = deser_cql_value(typ, &mut val).map_err(|err| {
```
Before deser refactor all deserialization was going trough `deser_cql_value` and all actual deserialization logic was there, right?
IIUC after you add `DeserializeValue` implementation the deserialization logic will be duplicated, right?
I think `deser_cql_value` should use `DeserializeValue` somehow and not duplicate logic.
Maybe it's addressed in later commits, if so feel free to ignore.

Lorak-mmk replied:
This would also allow us to dump `GenericParseError`, right?

wprzytula replied:
> This would also allow us to dump `GenericParseError`, right?

Not really, unfortunately, as some low-level deserialization error (think `types` module) are represented using `ParseError`.

ATM `ParseError` has a variant of `DeserializationError`, and `DeserializationError` can contain `ParseError` inside... That's a mess, but an errors refactor is needed here to somehow split `ParseError` into finer-grained pieces. E.g. `LowLevelDeserializationError`, which would be returned from `types` module routines and it would fall into `DeserializationError`.

wprzytula replied:
> I think deser_cql_value should use DeserializeValue somehow and not duplicate logic.

In new two commits, I've refactored `deser_cql_value` and `deser_rows` to use the new framework, so much less duplication and better error messages.

Lorak-mmk commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -73,6 +73,49 @@ where
 
+/// Values that may be empty or not.
+///
+/// In CQL, some types can have a special value of "empty", represented as
+/// a serialized value of length 0. An example of this are integral types:
+/// the "int" type can actually hold 2^32 + 1 possible values because of this
+/// quirk. Note that this is distinct from being NULL.
```
If it's distinct from null  doesn't it mean that int can hold 2^32 + 2 values (int values + null + empty)?

Lorak-mmk commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -984,6 +984,70 @@ where
+    }
+}
+
+macro_rules! impl_tuple_multiple {
+    (;< $($T_PREF:ident,)* >;;< $($idx_pref:literal,)* >;;< $($idf_pref:ident,)* >) => {
+        impl_tuple!(;;);
+    };
+    ($TN:ident $(,$Ti:ident)*; < $($T_PREF:ident,)* > ;  $idx_n:literal $(,$idx:literal)*; < $($idx_pref:literal,)* >; $idf_n:ident $(,$idf:ident)* ; <$($idf_pref:ident,)*>) => {
+        impl_tuple!($($T_PREF,)* $TN ; $($idx_pref, )* $idx_n; $($idf_pref, )* $idf_n);
+        impl_tuple_multiple!($($Ti),* ; < $($T_PREF,)* $TN, > ; $($idx),*; < $($idx_pref, )* $idx_n,  >; $($idf),*; < $($idf_pref, )* $idf_n, >);
+    }
+}
```
This macro needs some explanation about how it works, it probably won't be easy to understand for a new contributor.

wprzytula replied:
For writing a comment explaining how it works, first I have to remember and understand it... :P

Lorak-mmk commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -122,6 +123,96 @@ impl<'frame> DeserializeRow<'frame> for ColumnIterator<'frame> {
+    }
+}
+
+macro_rules! impl_tuple_multiple {
+    (;< $($T_PREF:ident,)* >;;< $($idx_pref:literal,)* >;;< $($idf_pref:ident,)* >) => {
+        impl_tuple!(;;);
+    };
+    ($TN:ident $(,$Ti:ident)*; < $($T_PREF:ident,)* > ;  $idx_n:literal $(,$idx:literal)*; < $($idx_pref:literal,)* >; $idf_n:ident $(,$idf:ident)* ; <$($idf_pref:ident,)*>) => {
+        impl_tuple!($($T_PREF,)* $TN ; $($idx_pref, )* $idx_n; $($idf_pref, )* $idf_n);
+        impl_tuple_multiple!($($Ti),* ; < $($T_PREF,)* $TN, > ; $($idx),*; < $($idx_pref, )* $idx_n,  >; $($idf),*; < $($idf_pref, )* $idf_n, >);
+    }
+}
```
Again, some comment would be nice

wprzytula replied:
What kind of a comment would you like here?
What comes to my mind is "Macro to implement tuple deserialization for various sizes of tuple", which is not much helpful.

wprzytula commented on scylla-cql/src/types/deserialize/frame_slice.rs:
```diff
@@ -0,0 +1,188 @@
+            frame_subslice: mem,
+            original_frame: frame,
+        }
+    }
```
> Other methods guarantee that `frame_subslice` points to the same slice as `original_frame`, but this one allows creation of ill-formed objects.

This is not necessarily true. Consider `as_slice_mut`. The caller of it can even replace the slice with an arbitrary slice satisfying the given lifetime - e.g. using a `'static` slice.

> Modifying `FrameSlice` using `as_slice_mut` seems less error-prone to me than creating new slice and then a new object using this slice.

Mostly yes.

Pushed a new version with `new_subslice` ditched.

wprzytula commented:
v4.3:
- ditched `FrameSlice::new_subslice()`;
- fixed `read_cql_bytes()` so that they really satisfy their guarantee about not mutating the FrameSlice in case of an error.

wprzytula commented on scylla-cql/src/types/deserialize/frame_slice.rs:
```diff
@@ -0,0 +1,188 @@
+            Ok(None) => Ok(None),
+            Err(err) => Err(err),
+        }
+    }
```
I've rewritten it in a different way, to really satisfy the given guarantee:
`/// If the operation fails then the slice remains unchanged.`

wprzytula commented on scylla-cql/src/types/deserialize/row.rs:
```diff
@@ -0,0 +1,85 @@
+/// one of the types having support built into the driver.
+#[derive(Debug, Error, Clone)]
+#[error("Failed to type check column types against the Rust type {rust_name}: {kind}")]
+pub struct BuiltinTypeCheckError {
+    /// Name of the Rust type used to represent the values.
+    pub rust_name: &'static str,
+
+    /// Detailed information about the failure.
+    pub kind: BuiltinTypeCheckErrorKind,
+}
```
I based this error on `serialize::row::BuiltinTypeCheckError`, which lacks `cql_names` too.

wprzytula commented:
v4.4:
- added `got` field to `row::BuiltinTypeCheckError`, so that error message is richer in CQL column types now. Adjusted tests to check that.

wprzytula commented:
v4.4:
- fixed bad tuple index in `DeserializeValue` impl for tuples.

wprzytula commented:
v4.5:
- corrected docstring on `FrameSlice::read_cql_bytes()`,
- renamed `TupleDeserializationError::ElementDeserializationFailed` to `FieldDeserializationFailed`,
- added `TupleTypeCheckError::FieldDeserializationFailed`,
- fixed tuple `DeserializeValue` impl to correctly wrap errors,
- added tests for tuple errors (passing only after the above changes).

wprzytula commented:
v5.0:
- refactored `deser_cql_value` and `deser_rows` to use the new deserialization framework. Wins:
  1. less code duplication,
  2. better error messages in the old framework (coming from the new framework).

wprzytula commented on scylla-cql/src/types/typecheck_errors/mod.rs:
```diff
@@ -1,2 +1,33 @@
+///   but it might be nested in the [`row_errors::BuiltinTypeCheckError`].
+#[derive(Debug, Clone, Error)]
+#[error(transparent)]
+pub struct TypeCheckError(pub(crate) Arc<dyn std::error::Error + Send + Sync>);
```
This is an attempt to make type checking and deserialization more type safe wrt errors.

wprzytula commented:
> v5.0:
> 
>     * refactored `deser_cql_value` and `deser_rows` to use the new deserialization framework. Wins:
>       
>       1. less code duplication,
>       2. better error messages in the old framework (coming from the new framework).

Keep in mind that now, with those changes introduces, compatibility tests between the new and the old framework and no longer relevant, because the old framework's backbone (`deser_cql_value` with all the impls) is deleted. Then:
1. Should we even keep those compatibility tests? They no longer prove anything. After all, they proved what we wanted them to prove, in earlier commits, and now are useless.
2. Should we perhaps postpone introducing those changes until we delete the whole old framework altogether? Then the tests are still valid.

wprzytula commented:
v5.2: `impl_tuple_multiple()` macro related changes:
- documented the macro;
- added a special entry rule to the macro, so now its syntax is cleaner (no longer requires `<`, `>` characters to be passed) **and** the impl for `()` comes first, not last; 
- deduplicated the macro by importing it to `row.rs` from `value.rs`.

wprzytula commented:
v5.3:
- rebased on main.

wprzytula commented:
v5.3.1:
- fixes after rebasing on main.

wprzytula commented:
v5.4:
- added a convenience function to be used in new tests (not yet pushed)

wprzytula commented:
v6.0:
- removed old/new framework compatibility tests, as they no longer test anything;
- wrote a suite of ser/de identity tests; they assert that a value after serialization and then deserialization is the same as before.

wprzytula commented:
v6.0.1:
- fixed one import not found for MSRV Rust

wprzytula commented:
v6.0.2:
- fixed datetime tests

Lorak-mmk commented on scylla-cql/src/types/deserialize/result.rs:
```diff
@@ -0,0 +1,195 @@
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, Some(self.remaining))
```
This 0 should be self.remaining. Iterator only returns `None` after `self.remaining` becomes 0, which happends precisely after `self.remaining` iterations

muzarski commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -628,6 +632,195 @@ where
+}
+
+impl<'frame, T> ListlikeIterator<'frame, T> {
+    pub fn new(
```
ditto

muzarski reviewed (commented):
Just 3 nitpicks, otherwise LGTM.

wprzytula commented:
Resolved comments:
- fixed `size_hint`,
- tuple `element` -> `field`.

wprzytula commented:
v6.0.5:
- decreased `pub` visibility of some functions

wprzytula commented:
v6.0.6:
- sprinkled `#[inline]` over short `pub` functions.

wprzytula commented:
v6.0.7:
- renamed `BuiltinTypeCheckErrorKind::ColumnTypeMismatch` to `ColumnTypeCheckFailed`

wprzytula commented:
v6.0.8:
- removed `exp()` test, used for review purposes only.

wprzytula commented:
v6.1:
- constrained pub reexports in scylla/lib.rs.

Lorak-mmk commented:
> v6.1:
> 
>     * constrained pub reexports in scylla/lib.rs.

I don't see a push for this change. Did you forget to push it or Github broke displaying somehow?

wprzytula commented:
> > v6.1:
> > ```
> > * constrained pub reexports in scylla/lib.rs.
> > ```
> 
> I don't see a push for this change. Did you forget to push it or Github broke displaying somehow?

If you look above, you'll see [scylla lib: constrain pub deserialize reexports](https://github.com/scylladb/scylla-rust-driver/pull/970/commits/00859a956a7723e5c4e24336dd7a4eeb1dc9f7ad). The reason why GH didn't show it is that I added it as a new commit on top of others.

wprzytula commented:
v6.2:
- fixes to docstrings.

muzarski reviewed (approved):
Please rebase on top of main so the CI passes.

wprzytula commented:
Rebased on main.

wprzytula commented:
v6.3:
- changed UdtIterator Item a bit (will be useful for derive macros in the follow up)
- added test cases that checks errors when collections are attempted to be deserialized from null.
